File: Python3/_28.py
class Solution:
    def strStr(self, hay: str, needle: str) -> int:
        if not needle:
            return 0
        elif len(hay) < len(needle):
            return -1

        # Keep 2 points (start, end) for hay and 1 for needle
        iNEnd = 0
        iHStart = 0
        iHEnd = 0
        while (iNEnd < len(needle) and iHEnd < len(hay)):

            # If find match, ++ current ends in both strings
            if needle[iNEnd] == hay[iHEnd]:
                iHEnd += 1
                iNEnd += 1

            # Reset needle to none found and ++ iHstart because
            # we just checked starting from that in hay
            else:
                iNEnd = 0
                iHStart += 1
                iHEnd = iHStart

        # Finished checking needle (success) or hay (failure)
        if iNEnd == len(needle):
            return iHStart;
        else:
            return -1

'''
# List comprehension makes sliding window easier and quicker
class Solution:
    def strStr(self, hay: str, needle: str) -> int:

        if not needle:
            return 0
        elif len(needle) > len(hay):
            return -1

        needleLen = len(needle)

        # Consider only chars in hay that if is start, matching
        # end dependent on needleLen remains within hay
        for i in range(len(hay) - needleLen + 1):
            # Slide window of length needleLen in hay
            if hay[i:i + needleLen] == needle:
                return i

        return -1
'''

File: Java/_28.java
class Solution {
    public int strStr(String hay, String needle) {
    	if (needle.length() == 0) return 0;
    	else if (hay.length() < needle.length()) return -1;

    	int iNeedleEnd = 0;
    	int iHayStart = 0;
    	int iHayEnd = 0;

    	while (iNeedleEnd < needle.length() && iHayEnd < hay.length()) {
    		if (needle.charAt(iNeedleEnd) == hay.charAt(iHayEnd)) {
    			iNeedleEnd++;
    			iHayEnd++;
    		}
    		else {
    			iNeedleEnd = 0;
    			iHayStart++;
    			iHayEnd = iHayStart;
    		}
    	}

    	if (iNeedleEnd == needle.length()) return iHayStart;
    	else return -1;
    }
}